chain/chain/src/migrations.rs

```diff
@@ -0,0 +1,41 @@
+use crate::store::ChainStoreAccess;
+use crate::types::RuntimeAdapter;
+use near_chain_primitives::error::Error;
+use near_primitives::hash::CryptoHash;
+use near_primitives::types::ShardId;
+
+/// Check that epoch of block with given prev_block_hash is the first one with current protocol version.
+fn is_first_epoch_with_protocol_version(
+    runtime_adapter: &dyn RuntimeAdapter,
+    prev_block_hash: &CryptoHash,
+) -> Result<bool, Error> {
+    match runtime_adapter.get_prev_epoch_id_from_prev_block(prev_block_hash) {
+        Ok(prev_epoch_id) => {
+            let epoch_id = runtime_adapter.get_epoch_id_from_prev_block(prev_block_hash)?;
+            let protocol_version = runtime_adapter.get_epoch_protocol_version(&epoch_id)?;
+            let prev_epoch_protocol_version =
+                runtime_adapter.get_epoch_protocol_version(&prev_epoch_id)?;
+            Ok(protocol_version != prev_epoch_protocol_version)
+        }
+        Err(_) => Ok(true),
```

Comment:
This looks very suspicious to me. Have you accounted for all possible ways an attacker can maliciously send you some block that such `get_prev_epoch_id_from_prev_block` returns an error?

Reply:
I've tried to handle error coming from `KeyValueRuntime.get_prev_epoch_id_from_prev_block` but now found that my implementation was incorrect. Fixed